AC-414 adding headings as first child to sections

clrux opened the pull request:
# [AC-414](https://openedx.atlassian.net/browse/AC-414)

WCAG spec indicates that a heading should be the first child of any `section` element. Furthermore, the section should be labeled by the heading to identify sections when browsed via landmarks.

This addresses the Instructor dashboard pages and sections and adds dynamically built headings and labels to satisfy this requirement.

I've also added heading classnames which will eventually sync up with the UXPL for styling. I also opened up the UXPL heading styles to be available anywhere in LMS so these new classes can come into play. If no UXPL classnames are supplied on LMS headings, they'll fall back to what they've always looked like so there shouldn't be any visual change beyond the scope of this work.

I've also corrected some heading levels so an appropriate level is used where needed. Previously it seemed as though a heading level was chosen based on how it appeared visually, which isn't uncommon.

## Notes

There are a number of a11y test failures that aren't included in the scope of this PR. I'm ignoring the irrelevant tests, but I've added a TODO comment with the tag "wcag2aa" so we can find these later.

## Todos

- [x] Fix sections and headings
- [x] Add basic a11y tests
- [x] Make templates safe
- [x] Create [custom a11y rules/checks for section + heading markup](https://github.com/edx/edx-custom-a11y-rules/pull/3)

## Sandbox

https://clrux-ac-414.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/info

## Reviewers

- [ ] @benpatterson   
- [x] @cptvitamin 
- [x] @cahrens

clrux commented:
jenkins run lettuce

benpatterson commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -21,7 +20,7 @@
   <div class="reports-download-container action-type-container">
     <hr>
-    <h2> ${_("Reports")}</h2>
+    <h3 class="hd hd-2">${_("Reports")}</h3>
```
why change h2 to h3?

clrux replied:
I had to update the headings to follow proper hierarchy. There's an H2 on line 4, which represents the content of the section, so the next logical heading level is 3.

benpatterson commented on lms/templates/instructor/instructor_dashboard_2/edit_coupon_modal.html:
```diff
@@ -3,7 +3,8 @@
 %>
-<section id="edit-coupon-modal" class="modal" role="dialog" tabindex="-1" aria-label="${_('Edit Coupon Code')}">
+<section id="edit-coupon-modal" class="modal" role="dialog" tabindex="-1" aria-labelledby="header-edit-coupon-code">
+    <h2 id="header-edit-coupon-code">${_("Edit Coupon Code")}</h2>
```
why not `modal` in the `id` attribute?

benpatterson replied:
i also see repetition of the id with the `aria-labelledby`. can this be more programmatic?

clrux replied:
I didn't include "modal" in the heading `id` value because while this header is part of the modal container, it itself isn't representative of a modal. If anything, this may be updated to `heading-edit-coupon-code` or `edit-coupon-code-heading`. Would either of those be preferable for you?

benpatterson commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -33,7 +33,7 @@
 
 <div class="batch-enrollment membership-section">
-  <h2> ${_("Batch Enrollment")} </h2>
+  <h3 class="hd hd-2">${_("Batch Enrollment")}</h3>
```
Looking at your sandbox, I think you need some additional differentiation between h3 and what follows it. That is to say, I expect more kerning or space, etc after "Batch Enrollment". It looks compressed to me.

clrux replied:
Fair point. My goal here was to focus on functionality and markup, not styles, but I can see the value in updating the styles too.

A few months ago I brought over UXPL styles to handle these new heading classes, but scoped them a specific container. I might make another PR that opens them up and makes them available to a larger scope to cover these instances. Would you be OK with a second follow-up PR for that, or would you prefer I get them into this PR?

benpatterson replied:
Hmm that sounds like a lot of scope creep to me. I'd fix it here and mark a TODO for adjusting the style once it's been adjusted.

benpatterson replied:
By the way, when you say your goal is to focus on functionality and markup, that does not mean that you ignore the styling implications, and when introducing text such as a header, it's important to understand the impact.

benpatterson commented on lms/templates/instructor/instructor_dashboard_2/student_admin.html:
```diff
@@ -1,10 +1,10 @@
   %if section_data['is_small_course']:
   ## Show the gradebook for small courses
-    <h2>${_("Student Gradebook")}</h2>
+    <h3 class="hd hd-2">${_("Student Gradebook")}</h3>
```
Same comment here. h3 is looking compressed with the text that follows.
![image](https://cloud.githubusercontent.com/assets/2004678/15298704/991a4220-1b6e-11e6-9fb3-0b45719908db.png)

clrux replied:
@benpatterson  See https://github.com/edx/edx-platform/pull/12452#discussion_r63395424 Thoughts?

benpatterson commented:
We should see tests checked in with this change to ensure no one introduces the anti-pattern later.

clrux commented:
@benpatterson I talked with Andy about the heading styles and he suggested I just go ahead and make the [previously scoped, new classnames](https://github.com/edx/edx-platform/pull/11764/files) available to the entire LMS. [This commit](https://github.com/edx/edx-platform/pull/12452/commits/4203396ce49775284134de5fffe9b93dc40c18a3) does that relatively quickly. I made it as a separate commit in case you still think I should break this out into it's own PR, but the change was small enough I thought it might be OK to get into this PR so the changes are all self-contained.

Sandbox has been updated for preview.

andy-armstrong commented on lms/static/sass/xmodule/_headings.scss:
```diff
@@ -88,29 +88,33 @@ $headings-base-color:                       $gray-d2;
-.xblock .xblock {
+// let's make these classnames available to the entire LMS!
+
+.hd-1,
```
Can you move these classes to the following file as they are no longer scoped to XBlocks:

https://github.com/edx/edx-platform/blob/master/lms/static/sass/base/_base.scss

clrux replied:
I could. Do you think it's more helpful to have them in their own file though? Since these are duplicates of whatever we have set in the UXPL, if the UXPL changes, we'd need to update these accordingly and I feel like having them in their own file is easier to manage. What do you think?

andy-armstrong replied:
It makes sense to me to introduce a new file for Pattern Library compatibility if you want to do that. I just don't want to have global styles defined beneath xmodule because nobody will expect to find them there.

clrux replied:
Ah yes, the `xmodule` directory is problematic. How about a `_headings.scss` partial in the main directory (alongside `_base.scss`, and the other extends and mixins)?

andy-armstrong replied:
Sounds good.

clrux replied:
Done.

clrux commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -1154,3 +1154,29 @@ def test_user_not_enrolled_error(self):
+        # before 2 October 2016, and include the below exclusions.
+        self.instructor_dashboard_page.a11y_audit.config.set_scope(
+            include=[],
+            exclude=["table", "a"]
```
These failures aren't included in the scope of this work, so I'm excluding them from running. However, we should tackle these soon to ensure proper accessibility on this page.

clrux commented on lms/static/js/spec/main.js:
```diff
@@ -703,6 +708,10 @@
+        'js/spec/shoppingcart/shoppingcart_spec.js',
+        'js/spec/instructor_dashboard/instructor_dashboard_spec.js',
+        'js/spec/instructor_dashboard/ecommerce_spec.js',
+        'js/spec/instructor_dashboard/student_admin_spec.js',
```
The only thing added here is the instructor dashboard spec. I'm not sure why the other three are showing up as added. I did have to rebase, but this should be comparing to master which should already have them?

cahrens replied:
They do exist, but are further down in the file. So basically, you duplicated them.

In any case, your changes to this file will be reverted as part of deleting the unit test.

clrux commented on lms/static/sass/base/_headings.scss:
```diff
@@ -88,29 +88,33 @@ $headings-base-color:                       $gray-d2;
-            @extend %hd-#{$i};
-        }
-    }
+.xblock .xblock {
```
These H3s are funky and some of them need to be scoped to xblocks, we I'm keeping this particular H3 scoped.

cahrens replied:
Why is the xblock class repeated (.xbock .xblock)?

clrux replied:
That's how the nesting works in the markup :(

clrux commented on lms/templates/ccx/coach_dashboard.html:
```diff
@@ -68,24 +68,24 @@
             </li>
           </ul>
-          <section id="membership" class="idash-section">
+          <section id="membership" class="idash-section" aria-label="${_("CCX Membership Management")}">
```
Because the templates for each section are dynamically loaded, there's no guarantee that the IDs for the headings in those files will stay the same. I'm using a fixed aria-label here - and for the others - so that content within each included file may be flexible.

clrux commented on lms/templates/instructor/instructor_dashboard_2/add_coupon_modal.html:
```diff
@@ -3,7 +3,8 @@
 from django.core.urlresolvers import reverse
 %>
-<section id="add-coupon-modal" class="modal" role="dialog" tabindex="-1" aria-label="${_('Add Coupon Code')}">
+<section id="add-coupon-modal" class="modal" role="dialog" tabindex="-1" aria-labelledby="header-coupon-code">
```
Labelled by the child heading `id`. These sections aren't dynamically generated on this page, so there currently isn't a way to make this programatic. Each `section` here should be labelled by its child heading.

clrux commented on lms/templates/instructor/instructor_dashboard_2/course_info.html:
```diff
@@ -35,7 +34,7 @@
 
 <div class="basic-wrapper">
-  <h2>${_("Basic Course Information")}</h2>
+  <h3 class="hd hd-3">${_("Basic Course Information")}</h3>
```
Using an H3 here because it's next in the structure hierarchy.

clrux commented:
jenkins run js

benpatterson commented:
I understand the reasons behind making changes in the underlying structures, but we are now touching more universally-used areas. I'm not stopping this PR...I recognize you guys have collaborated on it and again the reasons why. I just want to be sure this is carefully tested before it's merged.

cahrens commented:
Since this isn't video-related, I don't think it falls to @muzaffaryousaf. I can review it when it is ready.

clrux commented:
@muzaffaryousaf Was your thumb on Christina's comment that you "approve" this PR, or are you agreeing with her? If the latter, it's ready :)

clrux commented:
@benpatterson As for "carefully tested" could you clarify? The work here, save the Sass, is pretty tightly scoped to the instructor dashboard and deals less with functionality and more with proper markup across the dashboard templates. There's a JS test here to ensure this structure is present, and I've manually verified the instructor dashboard pages (also verifiable on the sandbox).

clrux commented:
@cahrens This is ready for your review.

cahrens commented:
@clrux Great-- I will try to get to it today.

cahrens commented on lms/static/js/spec/instructor_dashboard/instructor_dashboard_spec.js:
```diff
@@ -0,0 +1,22 @@
+        describe("LMS Instructor Dashboard", function() {
+            
+            beforeEach(function() {
+                loadFixtures('js/fixtures/instructor_dashboard/instructor_dashboard.html');
```
Because this test is simply checking the structure of a fixture file, it doesn't have any value. There is no way to enforce that the fixture file is the same as the HTML actually returned on the server. I think you should delete it and the fixture file.

Fixture files should only be used for testing JavaScript interaction (in which case you do need the HTML elements present, and the actual structure of the HTML is not that important).

clrux replied:
This test will be useful for AC-415 so I will move these files there and have them test the functionality of the new tabbed navigation. Thanks!

cahrens commented:
@clrux I'm getting a 500 error on https://clrux-ac-414.sandbox.edx.org/. Can you get the sandbox working?

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -1154,3 +1156,29 @@ def test_user_not_enrolled_error(self):
+        # a11y test added on 17 May 2016, but mostly disabled as there's a slew
+        # of issues on the Instructor Dashboard. We should aim to resolve these
+        # before 2 October 2016, and include the below exclusions.
+        self.instructor_dashboard_page.a11y_audit.config.set_scope(
```
Is there no value in disabling the individual checkpoints that fail? Is it so bad that nothing would actually be tested anyway?

cahrens replied:
Also, the instructor dashboard is made up of many "tabs". Unless all the HTML for all the tabs is somehow getting checked here, I would recommend adding individual test_a11y test points for the various tabs of the instructor dashboard (hopefully some have less egregious errors than others).

With this approach, I wouldn't create a separate LMSInstructorDashboardA11yTest test, but instead add a test point to the various subclasses BaseInstructorDashboardTest, assuming that they represent testing of the individual "tabs".

clrux replied:
Alternatively, I can add this as part of AC-415 which fixes probably 90% of the found accessibility issues.

cahrens commented:
@clrux I've looked through the PR and noted some issues with testing. The HTML itself looks fine to me, but I want to be able to manually test it on the sandbox.

clrux commented:
@cahrens Sandbox is refreshed. Going through your initial feedback now.

cahrens commented:
@clrux and I chatted some about adding automated tests for the HTML structure changes in this PR. I don't think it is worth it in this case-- we don't test the actual structure of an HTML file, we instead test the behavior. JavaScript unit tests are useless in this case, because they would have to use fixtures (since most of the actual HTML is rendered server-side), and verifying the structure of a fixture file does not enforce that the correct structure exists in the production HTML.

Testing through Python unit tests or Bok Choy would require many different tests/scenarios to cover all the HTML files. Again, we have not invested in testing at this level, and I'm not convinced that it would be particularly helpful.

clrux commented:
@cahrens I think @benpatterson is back on Monday so I'll confirm the absence of tests for this PR is OK (there will be some added in [follow-up work](https://github.com/edx/edx-platform/pull/12514)).

cahrens commented:
Is it necessary to add these very large headers? This looks odd to me.

![image](https://cloud.githubusercontent.com/assets/484484/15405065/5a95c3f0-1dce-11e6-9e32-5b49a617e23e.png)

And it messes up the peer relationship. For example, on this tab "Data Download" was intended to be a peer to "Reports".

![image](https://cloud.githubusercontent.com/assets/484484/15405154/bbdc4ba2-1dce-11e6-9f75-23397aab3d0c.png)


Note that there isn't one on the Analytics tab.

clrux commented:
@cahrens We are incorporating the UXPL heading styles here, so the way they look will likely change. I will verify which are children and which are siblings. There also seem to be many other options for tabs here than originally planned, with an unlimited number possible. I may re-think the way this is done.

clrux commented:
Please hold off on reviewing further. I'm investigating a possible simpler solution to this PR. I will ping folks for a second pass when I'm ready.

clrux commented on lms/static/sass/_build-course.scss:
```diff
@@ -4,6 +4,7 @@
 // About: Sass compile for the LMS Courseware Elements that are shared between LTR and RTL UI. Configuration and vendor specific imports happen before this shared set of imports are compiled in the lms-course-*.scss files.
 
 @import 'base/base';
+@import 'base/headings';
```
Including the new Sass partial for heading styles.

clrux commented on lms/templates/instructor/instructor_dashboard_2/cohort-editor.underscore:
```diff
@@ -9,7 +9,7 @@
         <form action="" method="post" id="cohort-management-group-add-form" class="cohort-management-group-add-form">
 
-            <h4 class="form-title"><%- gettext('Add students to this cohort') %></h4>
+            <h4 class="hd hd-3 form-title"><%- gettext('Add students to this cohort') %></h4>
```
Semantically an H4, but it looks like an H3.

clrux commented on lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html:
```diff
@@ -116,11 +116,12 @@
-          <section id="${ section_data['section_key'] }" class="idash-section">
-            <%include file="${ section_data['section_key'] }.html" args="section_data=section_data" />
+          <section id="${ section_data['section_key'] }" class="idash-section" aria-labelledby="header-${section_data['section_key']}">
+              <h2 class="hd hd-2 sr-only" id="${ section_data['section_key'] }">${ section_data['section_display_name'] }</h2>
```
Ensuring a proper H2 on every dynamically loaded template.

clrux commented:
Hi @benpatterson, @cptvitamin, and @cahrens this is ready for your review. Let me know if you have questions. The sandbox has been updated.

cptvitamin commented:
@cahrens I disagree that Data Download is meant to be a peer/sibling to Reports and the other headings on that tab.  It looks to me like the very first section is missing a heading that would introduce it.  Data Download, as poor as it is, is a title for the entire tab.

cptvitamin commented:
@clrux get rid of the sr-only on the h2's and this is good to go from my perspective.

benpatterson commented:
You should address the bokchoy failures. 

When putting in an updated pull request, next time could you wait to squash? It makes it easier on reviewers.

clrux commented:
jenkins run bokchoy

clrux commented:
@benpatterson Yeah, I got squash happy :(

cahrens commented:
Unfortunately there are 53 violations in these files from the safe commit linter.

https://build.testeng.edx.org/job/edx-platform-quality-pr/18731/artifact/edx-platform/reports/safecommit/safecommit.report/*view*/

This is a pretty extreme case, and I wouldn't recommend you try to fix all these issues as part of this PR (FYI @robrap). But I wanted to mention it in case you aren't in the habit yet of checking these new quality reports.

clrux commented:
Either Github is still running slow or you're working super late :) A lot (most) of these safe template issues look to be pre-existing, but triggered maybe because the files were changed? I'm happy to address them - a good learning experience - but maybe in another PR. Would that work?

cahrens commented on lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html:
```diff
@@ -116,11 +116,12 @@
         % for section_data in sections:
-          <section id="${ section_data['section_key'] }" class="idash-section">
-            <%include file="${ section_data['section_key'] }.html" args="section_data=section_data" />
+          <section id="${ section_data['section_key'] }" class="idash-section" aria-labelledby="header-${section_data['section_key']}">
```
The aria-labelledby doesn't match the ID, does it? Is this something the bok choy accessibility test should catch?

clrux replied:
Hm, good catch. I've done so much back and forth with these files, it looks like the `header-` prefix never made it back in. Yes, the `aria-labelledby` here should match the `id` of the child heading. I had it set up to programmatically make this happen, but in all the shuffle, it looks like it was lost. I will get this back in there.

FWIW, the whole instructor dashboard is a tangle and should probably be re-built at some point. Originally I had modified way more files than I needed to, then simplified it to just the base six or so files that are included, then realized that certain settings differ from localhost and a sandbox (e.g., Analytics, etc.) so needed to go through them manually using the `Sections` definition in the main Python loader file. There's gotta be a simpler way to do all of this.

clrux replied:
I don't think the accessibility checks test to ensure things match, only that proper structure and valid attributes exist. It might be something I can add to a Jasmine test though - checking for matching values, etc. I started a Jasmine test for follow-up work in AC-415 (proper tab widgets) so I could include it there, or make a smaller test file here that is added to later in that work?

cahrens replied:
I think the bok choy test should catch it. I'd definitely go that route instead of Jasmine because we don't want to test fixture files for HTML structure.

clrux replied:
Yes, duh. We don't want to use prescribed "perfect" fixture markup. We want to use what's actually present in the real DOM. I'll take a look at the aXe documentation and see if this is something we can enable or possibly add.

The BokChoy test I wrote was moved to AC-415. I'd be curious to see what it finds here though, so maybe I'll move it back.

cahrens commented:
I'm done with my re-review. Looking good, just found one issue with a bad ID. But that makes me wonder-- what happened to the bok choy accessibility tests? I'd really like those to get in.

As far as the safe commit errors, the way it works is that the linter looks at all the files you touched, not just the lines you touched. This is because we want to improve the codebase pretty rapidly in this case. But I agree that it should be a separate PR if you take on some of that work, and you should check first with @robrap, as some of those files might be underway in other PRs.

benpatterson commented:
I agree we should have bok-choy a11y tests on the dashboard page. I didn't notice any in the test file, though I only took a quick look.

In terms of testing the markup, @clrux updated an already-existing unit test that looks at page elements. I assume he can add another test(s) that ensures any `<section>` is followed by some sort of `<h*>` ?

In terms of the safe linting, (a) I agree that's too much scope for this particular PR, and (b) a subsequent pr makes sense. @robrap or @cahrens how many fixes do you think would be reasonable, and could you priortize them for Chris. In other words, how many do you think he can do, and which ones? Two templates fair enough?

clrux commented:
@benpatterson @cahrens The latest commit fixes the earlier failing tests (turned out to be an issue with jQuery UI's accordion targeting an element that was updated) and includes an added test for the a11y stuff. I'm excluding `a` and `table` for now. The `a` will be fixed in AC-415. I've got the sandbox updating.

benpatterson commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -1154,3 +1156,29 @@ def test_user_not_enrolled_error(self):
+        # before 2 October 2016, and include the below exclusions.
+        self.instructor_dashboard_page.a11y_audit.config.set_scope(
+            include=[],
+            exclude=["a", "table"]
```
You should put a TODO comment in here, referencing the ticket that'll address it.

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -1154,3 +1156,29 @@ def test_user_not_enrolled_error(self):
+        # a11y test added on 24 May 2016, but mostly disabled as there's a slew
+        # of issues on the Instructor Dashboard. We should aim to resolve these
+        # before 2 October 2016, and include the below exclusions.
+        self.instructor_dashboard_page.a11y_audit.config.set_scope(
```
On a previous version of this PR, I asked a question about coverage with this test. Unfortunately I don't think I can get back the question, but I will try to recreate.

What exactly does this test cover? Will it cover tabs that aren't visible, or is it only covering the tab that is visible by default when you first visit the instructor dashboard? Assuming that it is only covering the tab that is visible by default, you should add separate a11y checks to the various subclasses of BaseInstructorDashboardTest so that each tab within the instructor dashboard is covered.

Also, did you verify if the aria-labelledby bug was caught by the acceptance test?

clrux replied:
This test covers all accessibility rules (excluding checks for `a` and `table`) within the Instructor Dashboard. I'm currently researching the breadth of tests aXe offers, and I don't think it checks that headings are immediate children of section elements. However, we do have a custom ruleset and I'm investigating adding this check to it (it's a separate repo).

clrux replied:
Upon further investigation, the W3 says headings _should_ be included within a section, not _must_ be included. I don't think we need to write a test to check for this, but should include documentation as a best practice.

For this PR, I've updated the markup structure to include such best practices and will include general level accessibility tests, but won't write any additional tests.

benpatterson replied:
Isn't the intention of this pull request to solve for that particular case? I don't understand why we'd put the time into fixing it, without putting in tests to assure it stays that way (that is, to assure that there are no regressions when folks are making other changes to this page, to the styling, etc). I do think it's a good idea to consider this for the custom aXe ruleset we use, but again we have unit tests that assert markup on this page. Can't we use that in this case?

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -1154,3 +1156,29 @@ def test_user_not_enrolled_error(self):
+    def setUp(self):
+        browser = os.environ.get('SELENIUM_BROWSER', 'firefox')
+
+        with patch.dict(os.environ, {'SELENIUM_BROWSER': browser}):
```
I'm not aware of any reason to do this (override SELENIUM_BROWSER) unless you are testing videos.

clrux replied:
👍

clrux commented:
Update for the day: I'm working on incorporating an a11y test into each class on the Instructor Dashboard - each section that is navigated to - rather than one large one that only checks the initial state. I'll ping again when I feel this is ready. Thanks for the input thus far.

robrap commented:
@clrux: I'm happy to sit with you if you have any questions, but I mostly documented what I was going to say.

Here are all the current violations:
https://build.testeng.edx.org/job/edx-platform-quality-pr/18813/artifact/edx-platform/reports/safecommit/safecommit.report/*view*/

I recommend you fix these first as part of this PR:
- lms/templates/file-upload.underscore
- lms/templates/instructor/instructor_dashboard_2/certificates.html
- lms/templates/instructor/instructor_dashboard_2/e-commerce.html (minor complexity in JavaScript)
- lms/templates/instructor/instructor_dashboard_2/notification.underscore

For other files, you can either fix all violations other than "mako-missing-default" to start, or choose some template(s) to fix completely, including "mako-missing-default".

Finally, I would leave instructor_dashboard_2/metrics.html for a separate ticket, because that's going to require the most work.  We could also try to determine if we could deprecate and remove this. :)

Here are the docs for fixing safe template linter violations:
edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter

clrux commented:
Thanks @robrap! I will begin to address these in this PR once I get the a11y tests implemented correctly.

clrux commented:
@cahrens I've added the individual BokChoy a11y tests. They all pass locally, but there are ignored tests in each. The `link-href` tests will be resolved in AC-415 but we should address the others before 2 October.

@benpatterson I talked to Mark this morning and while the W3 suggests using a heading as a child of a section, it's a suggestion at this point in time, not a requirement. However, it is good practice so I'll write tests to be included in our custom ruleset as a separate commit.

robrap commented on lms/templates/instructor/instructor_dashboard_2/certificates.html:
```diff
@@ -85,7 +85,7 @@ <h3 class="hd hd-3">${_("Generate Certificates")}</h3>
                 % else:
                     <p class="under-heading">
-                        ${_("When you are ready to generate certificates for your course, click Generate Certificates. You do not need to do this<br/>if you have set the certificate mode to on-demand generation.")}
+                        ${Text(_("When you are ready to generate certificates for your course, click Generate Certificates. You do not need to do this<br/>if you have set the certificate mode to on-demand generation."))}
```
You'll need to interpolate in `HTML(<br/>)` using a `format()` call.  I prefer when at least the format replacement variables are placed on the following line.

clrux replied:
Makes sense. FYI, this wasn't caught when I ran the `./scripts/safe_template_linter.py` locally.

clrux replied:
Maybe because HTML and Text are both valid?

robrap replied:
It's possible that the linter could fail on this, but there might be valid cases that would require disabling the linter.

The rule is, if you are in Mako and you have mixed HTML and plain text, wrap the translated string with `Text()` and interpolate in the `HTML(<tag>)` with `format()`.  Once you have this, it becomes pretty simple and consistent to fix issues.

robrap commented on lms/templates/instructor/instructor_dashboard_2/course_info.html:
```diff
@@ -97,9 +97,9 @@ <h3 class="hd hd-3">${_("Basic Course Information")}</h3>
-        ${_("View detailed Git import logs for this course {link_start}by clicking here{link_end}.").format(
-            link_start='<a href="{}">'.format(section_data['detailed_gitlogs_url']),
-            link_end='</a>'
+        ${HTML(_("View detailed Git import logs for this course {link_start}by clicking here{link_end}.")).format(
```
This top one should be Text().

robrap commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -96,7 +96,7 @@ <h3 class="hd hd-3">${_("Reports Available for Download")}</h3>
 
     ## Translators: a table of URL links to report files appears after this sentence.
-    <p>${_("<b>Note</b>: To keep student data secure, you cannot save or email these links for direct access. Copies of links expire within 5 minutes.")}</p><br>
+    <p>${HTML(_("<b>Note</b>: To keep student data secure, you cannot save or email these links for direct access. Copies of links expire within 5 minutes."))}</p><br>
```
This should use `Text()` with the HTML (e.g. `<b>`) interpolated in.  The reason for this is that it is possible that the translated text could have unexpected HTML characters in the translation that would need to be escaped.

robrap commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -118,7 +118,7 @@ <h3 class="hd hd-3">${_("Reports")}</h3>
 
             ## Translators: a table of URL links to report files appears after this sentence.
-            <p>${_("<b>Note</b>: To help protect learner data, links to these reports that you save outside of this page or that you send or receive in email expire after five minutes.")}</p><br>
+            <p>${Text(_("<b>Note</b>: To help protect learner data, links to these reports that you save outside of this page or that you send or receive in email expire after five minutes."))}</p><br>
```
Same as above.

robrap commented on lms/templates/instructor/instructor_dashboard_2/certificates.html:
```diff
@@ -112,7 +112,7 @@ <h3 class="hd hd-3"> ${_("Pending Tasks")} </h3>
         <form id="certificate-regenerating-form" method="post" action="${section_data['urls']['start_certificate_regeneration']}">
             <p class="under-heading">
-                ${_('To regenerate certificates for your course, choose the learners who will receive regenerated certificates and click <br/> Regenerate Certificates.')}
+                ${Text(_('To regenerate certificates for your course, choose the learners who will receive regenerated certificates and click <br/> Regenerate Certificates.'))}
```
Same as above.

robrap commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -265,12 +265,12 @@ <h3 class="hd hd-3">${_("Coupon Code List")}</h3>
              success: function (data) {
                  $('#error-msg').attr('class','success-msgs')
-                 $('#error-msg').html(data.message).show();
+                 $('#error-msg').text(data.message).show();
```
These all look good, but you'll need to test to be sure that data.message does not include any HTML tags.  If it does, you'll need to use HtmlUtils.  Same with the changes below.

robrap commented:
@clrux: Thanks for fixing these issues.  I left some comments.  Also, be sure to manually review, especially those messages that went from `html()` to `text()`.  I appreciate the help in chipping away at this issue.

clrux commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -97,9 +97,7 @@ <h3 class="hd hd-3">${_("Reports Available for Download")}</h3>
-        ${Text(_("{note_start}: To keep student data secure, you cannot save or email these links for direct access. Copies of links expire within 5 minutes.")).format(
-            note_start=HTML("<strong>Note</strong>")
-        )}
+        ${_("<strong>Note</strong>: To keep student data secure, you cannot save or email these links for direct access. Copies of links expire within 5 minutes.")}
```
@robrap Is this OK, given I added `expression_filter="h"` to the top of the file? It wasn't working the other way. Same with the below file.

robrap replied:
This seems like it shouldn't be working.  As written, this should show the text `<strong>Note</strong>: ...` as-is on the screen. 

The original was closer, but had an i18n issue.  I'm not sure what you mean by "it wasn't working".  I imagine this code should look as follows:
```
${Text(_("{strong_start}Note{strong_end}: To keep student data secure, you cannot save or email these links for direct access. Copies of links expire within 5 minutes.")).format(
    strong_start=HTML("<strong>"),
    strong_end=HTML("</strong>"),
)}
```
But again, I'm not sure what issue you were seeing.

clrux replied:
Oh, I was doing it slightly incorrectly. Thanks.

clrux commented:
Added rules/checks to our a11y repository: https://github.com/edx/edx-custom-a11y-rules/pull/3

clrux commented:
@benpatterson and @cahrens One more run-through when you get a chance, thanks. Ben, the tests that check for these changes are in the [other repo](https://github.com/edx/edx-custom-a11y-rules/pull/3) and will be picked up here, once those merge in.

cahrens commented:
@clrux To make this easier for me to re-review, can you squash all the commits since my last review? You could start the squash with "Updating a11y test removing browser update". If you'd like to keep a separate commit for safe template work, that's fine, but there seemed to be a lot of tweaking in subsequent commits, so I'm having trouble reviewing.

Thanks!

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -361,6 +391,84 @@ def test_can_reset_attempts(self):
+        """
+        Proctored: Track selection accessibility tests
+        """
+        self.track_selection_page.a11y_audit.config.set_rules({
```
What is actually visible when this test is run? Unlike previous classes, the setUp for this class does not appear to visit the part of the instructor dashboard related to the test. You need to log in and visit the appropriate place first.

clrux replied:
FWIW these work for me in their current state. The page object is established and available already. The `setUp` methods in other classes are when there's a separate class for the accessibility tests. Since we're including these checks within classes, we don't need to perform the set up again.

I was getting failures locally and on Jenkins. I excluded the tests that were failing that aren't part of this PR and fixed the ones that were relevant.

cahrens replied:
The page object is created, but what is actually being viewed? The only "visit" in the setup that I see is 

```
# Add a verified mode to the course
         ModeCreationPage(
             self.browser, self.course_id, mode_slug=u'verified', mode_display_name=u'Verified Certificate',
             min_price=10, suggested_prices='10,20'
         ).visit()

```

Is that sufficient?

It is also my understanding that unless you use specific scoping for the a11y test, you test everything visible on the page. Therefore if these page objects are only regions within the same page (and they are all already visible), there is no need to run an accessibility check on every one-- it is just adding build time. 

That would also explain why they have all the same "ignores".

clrux replied:
You do bring up a good point in that I should limit the scope of the checks to the tab content area so that it's not repeating the full page check each time. Let me investigate.

I do see several `visit()`'s though, one for each section (in addition to the tab switch or section activation).

cahrens replied:
Can you stop by and show me where these visits are? This is the setUp method:

````
def setUp(self):
        super(ProctoredExamsTest, self).setUp()

        self.courseware_page = CoursewarePage(self.browser, self.course_id)

        self.course_outline = CourseOutlinePage(
            self.browser,
            self.course_info['org'],
            self.course_info['number'],
            self.course_info['run']
        )

        course_fixture = CourseFixture(**self.course_info)
        course_fixture.add_advanced_settings({
            "enable_proctored_exams": {"value": "true"}
        })

        course_fixture.add_children(
            XBlockFixtureDesc('chapter', 'Test Section 1').add_children(
                XBlockFixtureDesc('sequential', 'Test Subsection 1').add_children(
                    XBlockFixtureDesc('problem', 'Test Problem 1')
                )
            )
        ).install()

        self.track_selection_page = TrackSelectionPage(self.browser, self.course_id)
        self.payment_and_verification_flow = PaymentAndVerificationFlow(self.browser, self.course_id)
        self.immediate_verification_page = PaymentAndVerificationFlow(
            self.browser, self.course_id, entry_point='verify-now'
        )
        self.upgrade_page = PaymentAndVerificationFlow(self.browser, self.course_id, entry_point='upgrade')
        self.fake_payment_page = FakePaymentPage(self.browser, self.course_id)
        self.dashboard_page = DashboardPage(self.browser)
        self.problem_page = ProblemPage(self.browser)

        # Add a verified mode to the course
        ModeCreationPage(
            self.browser, self.course_id, mode_slug=u'verified', mode_display_name=u'Verified Certificate',
            min_price=10, suggested_prices='10,20'
        ).visit()

        # Auto-auth register for the course.
        self._auto_auth(self.USERNAME, self.EMAIL, False)

clrux replied:
Instructor Dashboard mixin (referenced in other classes): [L45](https://github.com/edx/edx-platform/blob/027ef8bfa731089d5f91cb99c8a80851a2256e18/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L45)
Bulk Email: [L59](https://github.com/edx/edx-platform/blob/027ef8bfa731089d5f91cb99c8a80851a2256e18/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L59)
Auto-enroll: [L94](https://github.com/edx/edx-platform/blob/027ef8bfa731089d5f91cb99c8a80851a2256e18/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L94)
Track selection: [L275](https://github.com/edx/edx-platform/blob/027ef8bfa731089d5f91cb99c8a80851a2256e18/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L275)
Verification flow: [L281](https://github.com/edx/edx-platform/blob/027ef8bfa731089d5f91cb99c8a80851a2256e18/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L281)
Immediate verification flow: might be able to get rid of this one?
Upgrade: might be able to get rid of this one?
Fake payment: might be able to get rid of this one?
Proctored: might be able to get rid of this one?
...
Taking a look at the others, but do the first ones here make sense?

cahrens replied:
OK, so I'm talking specifically about the a11y test points in this class. So anything in a different test class doesn't relate here. Let's look at test_track_selection_a11y-- you noted that the track selection "visit" method happens at L275. That's in the "_login_as_a_verified_user" helper method, so it only gets called if setUp or your test method calls something that calls that method. I don't believe that is the case for your a11y test method.

If this doesn't make sense, let's talk in person.

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -361,6 +391,84 @@ def test_can_reset_attempts(self):
+        self.track_selection_page.a11y_audit.check_for_accessibility_errors()
+
+    @attr('a11y')
+    def test_payment_verification_a11y(self):
```
Same thing with all of these-- you aren't visiting anything first.

clrux replied:
See comment above.

clrux commented:
@cahrens I squashed a few of the WIPs. Is this a bit better? I'm fighting the urge to squash too much because that seems to also be confusing to see.

cahrens commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -422,7 +422,7 @@ def _section_course_info(course, access):
     if settings.ANALYTICS_DASHBOARD_URL:
         dashboard_link = _get_dashboard_link(course_key)
-        message = _("Enrollment data is now available in {dashboard_link}.").format(dashboard_link=dashboard_link)
+        message = Text(_("Enrollment data is now available in {dashboard_link}.")).format(dashboard_link=dashboard_link)
```
I'm not clear what you are trying to do here. Should dashboard_link be marked as HTML? What is the motivation for this change?

clrux replied:
`_get_dashboard_link` brings back safe HTML. Previously, this was making this safe twice, which was spitting out HTML rather than the rendered HTML. Using `Text()` here ensures it uses the already safe HTML it receives from the `_get_dashboard_link` method.

cahrens replied:
HTML should not be wrapped with Text. @robrap can you provide some guidance here?

clrux replied:
I believe it was @robrap who helped me use the Text() approach.

robrap replied:
`Text()` is needed here because this is the equivalent to `Text(...).format(dashboard_link=HTML(...))`, only the `HTML` call is made earlier when creating dashboard_link.

cahrens replied:
@clrux can you please add a comment to the code, as this is confusing if you don't know the background.

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -361,6 +391,84 @@ def test_can_reset_attempts(self):
+        })
+        self.fake_payment_page.a11y_audit.check_for_accessibility_errors()
+
+    @attr('a11y')
```
Delete this test and self.problem_page-- I don't see that self.problem_page is ever referenced (certainly no "problem" is being visited).

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -361,6 +391,84 @@ def test_can_reset_attempts(self):
+        self.immediate_verification_page.a11y_audit.check_for_accessibility_errors()
+
+    @attr('a11y')
+    def test_upgrade_a11y(self):
```
This test class is... mysterious. I don't see that self.upgrade_page is actually used at all. Please delete it and the a11y test.

cahrens commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -361,6 +391,84 @@ def test_can_reset_attempts(self):
+        self.payment_and_verification_flow.a11y_audit.check_for_accessibility_errors()
+
+    @attr('a11y')
+    def test_immediate_verification_a11y(self):
```
This test class is... mysterious. I don't see that self. immediate_verification_page is actually used at all. Please delete it and the a11y test.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/certificates.html:
```diff
@@ -78,14 +79,14 @@
                 % else:
                     <p class="under-heading">
-                        ${_("When you are ready to generate certificates for your course, click Generate Certificates. You do not need to do this<br/>if you have set the certificate mode to on-demand generation.")}
+                        ${Text(_("When you are ready to generate certificates for your course, click Generate Certificates. You do not need to do this if you have set the certificate mode to on-demand generation."))}
```
There's no need to use "Text" here because you aren't mixing together Text and HTML.

robrap replied:
Agreed.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/certificates.html:
```diff
@@ -108,11 +109,11 @@
         <form id="certificate-regenerating-form" method="post" action="${section_data['urls']['start_certificate_regeneration']}">
             <p class="under-heading">
-                ${_('To regenerate certificates for your course, choose the learners who will receive regenerated certificates and click <br/> Regenerate Certificates.')}
+                ${Text(_('To regenerate certificates for your course, choose the learners who will receive regenerated certificates and click Regenerate Certificates.'))}
```
There's no need to use "Text" here because you aren't mixing together Text and HTML.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -256,7 +262,7 @@ <h3 class="hd hd-3">${_("Coupon Code List")}</h3>
          if($('#invoice_number').val() == "") {
                      $('#error-msg').attr('class','error-msgs')
-                     $('#error-msg').html("${_('The Invoice Number field cannot be empty.') | n, h, js_escaped_string}").show();
+                    $('#error-msg').text("${_('The Invoice Number field cannot be empty.') | n, js_escaped_string}").show();
```
@robrap I'm struggling to understand whether or not js_escaped_string is necessary here. Can you take a look and enlighten me?

cahrens commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -256,7 +262,7 @@ <h3 class="error-msgs" id="error-msg"></h3>
          if($('#invoice_number').val() == "") {
                      $('#error-msg').attr('class','error-msgs')
-                     $('#error-msg').html("${_('The Invoice Number field cannot be empty.') | n, h, js_escaped_string}").show();
+                    $('#error-msg').text("${_('The Invoice Number field cannot be empty.') | n, js_escaped_string}").show();
```
@robrap I'm struggling to understand whether or not js_escaped_string is necessary here. Can you take a look and enlighten me?

robrap replied:
It is needed.  The price we pay for embedding JavaScript in Mako that is more complex than just passing data through to a JavaScript file.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -265,12 +271,12 @@ <h3 class="error-msgs" id="error-msg"></h3>
              success: function (data) {
                  $('#error-msg').attr('class','success-msgs')
-                 $('#error-msg').html(data.message).show();
+                 $('#error-msg').edx.HtmlUtils.setHtml(data.message).show();
```
@clrux You have verified manually that this is working?

cahrens commented on lms/templates/instructor/instructor_dashboard_2/extensions.html:
```diff
@@ -1,8 +1,11 @@
+<%page args="section_data" expression_filter="h"/>
+<%!
+from django.utils.translation import ugettext as _
+from openedx.core.djangolib.markup import HTML, Text
```
Don't import things you aren't using. I don't see any references to HTML or Text.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html:
```diff
@@ -1,8 +1,10 @@
 <%!
 from django.utils.translation import ugettext as _
 from django.core.urlresolvers import reverse
+from openedx.core.djangolib.markup import HTML, Text
```
Don't import things you aren't using. I don't see any references to HTML or Text.

clrux commented:
jenkins run a11y

clrux commented:
@benpatterson and @cahrens One 'mo 'gain?

cahrens commented:
@clrux Can you squash together the commits since the last review so it is clear what I need to re-review? I'm working on the assumption that all commits on the branch will be squashed at the end, and therefore there is no need to keep small descriptive commits. If that is not your plan, we can discuss another way for me to get this information.

clrux commented:
@cahrens I believe you should be good from the "safe templates" commit.

clrux commented:
jenkins run bokchoy
jenkins run js

cahrens commented on lms/templates/instructor/instructor_dashboard_2/certificates.html:
```diff
@@ -4,6 +4,7 @@
 <%!
 from django.utils.translation import ugettext as _
 from openedx.core.djangolib.js_utils import dump_js_escaped_json, js_escaped_string
+from openedx.core.djangolib.markup import HTML, Text
```
Is the import needed? It's difficult for me to tell in this particular commit.

cahrens replied:
I don't believe this import is needed.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -271,12 +271,14 @@ <h3 class="hd hd-3">${_("Coupon Code List")}</h3>
              url: "${section_data['sale_validation_url'] | n, js_escaped_string}",
              success: function (data) {
                  $('#error-msg').attr('class','success-msgs')
```
Nit: instead of replicating the selector 3 times, I would suggest storing the JQuery element in a variable and then doing operations on it.

clrux replied:
I agree, but is this pre-existing code adjustment something that should be included in this PR? Happy to knock it out if so...

cahrens replied:
Nits are at your discretion-- just mentioning it as a general style guideline.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/send_email.html:
```diff
@@ -3,50 +3,48 @@
+                            %if cohort.name == section_data['default_cohort_name']:
+                            <br/>
+                            <div class="subheading">
+                                ${_("(Students without cohort assignment)")}
```
@efischer19 You removed "Students without cohort assignment" because it is incorrect, right? I don't understand why it is still in this PR.

efischer19 replied:
@catong I was intending to update this explanatory text, but missed it in my previous PR. What would be the best way to explain this? It's a cohort that may have been automatically created, if the following are all true:
- Cohorting is enabled for the course
- No automatic cohorts are created by staff
- A user without any cohort assignment access cohorted material

`Students without cohort assignment` is incorrect, because there may be users with a cohort of `None`, which is different. `Students with automatic assignment`, perhaps?
Further context: https://github.com/edx/edx-platform/pull/12602#discussion_r65573428

catong replied:
@efischer19 "Learners not explicitly assigned to a cohort" or "Learners not explicitly assigned to a named cohort"?

efischer19 replied:
@catong https://github.com/edx/edx-platform/pull/12695

cahrens commented on lms/templates/instructor/instructor_dashboard_2/send_email.html:
```diff
@@ -98,28 +95,18 @@
-            <div class="email-messages-wrapper"></div>
-        </div>
-        <div>
-            <p>${_("To see the status for all email tasks submitted for this course, click this button:")}</p>
```
@efischer19 and @clrux was this section (status for all email tasks submitted for this course) intentionally removed?

efischer19 replied:
This should still exist (unless @clrux intended to remove it), it's in master: https://github.com/edx/edx-platform/pull/12602#discussion_r65573428

cahrens commented:
@clrux The sandbox is not working. Please always update the sandbox with the latest version of the code and verify it is working before asking for a re-review.

cahrens commented:
The a11y bok choy tests look much better!

cahrens commented on lms/templates/ccx/coach_dashboard.html:
```diff
@@ -23,8 +23,8 @@
-        <section class="instructor-dashboard-content-2" id="ccx-coach-dashboard-content">
-          <h1>${_("CCX Coach Dashboard")}</h1>
+        <section class="instructor-dashboard-content-2" id="ccx-coach-dashboard-content" aria-labelledby="header-ccx-dashboard">
+          <h2 class="hd hd-2" id="header-ccx-dashboard">${_("CCX Coach Dashboard")}</h2>
```
This was previously an h1. Should it still be styled that way? I'm curious to see this page on the sandbox.

clrux replied:
I'd like to keep it consistent with other Instructor Dashboard-level sections. The logo of the page is the H1 so the next highest heading should be an H2.

cahrens replied:
OK

cahrens commented on lms/templates/instructor/instructor_dashboard_2/add_coupon_modal.html:
```diff
@@ -15,7 +16,7 @@
     <div id="coupon-content">
       <header>
-        <h2>${_("Add Coupon Code")}</h2>
+        <h3 class="hd hd-3">${_("Add Coupon Code")}</h3>
```
This previously was an h2. Should it still be styled that way? I don't know how to view this modal.

clrux replied:
See note above.

cahrens replied:
OK

clrux commented:
Spinning up another sandbox!

clrux commented:
@cahrens Sandbox ready. I've enabled Bulk Email and the CCX Dashboard.

pomegranited commented:
@clrux Looks like you're making great progress, but do you have an ETA on this change?  We're awaiting it so that our a11y checks can pass for PR #12582.

clrux commented:
@pomegranited Just waiting on reviewers to review. Hoping by Friday or Monday of next week?

robrap commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -265,12 +271,14 @@ <h3 class="error-msgs" id="error-msg"></h3>
                  $('#error-msg').attr('class','success-msgs')
-                 $('#error-msg').html(data.message).show();
+                 edx.HtmlUtils.setHtml($('#error-msg'), data.message);
+                 $('#error-msg').show();
```
@andy-armstrong: I think the following works:
```
edx.HtmlUtils.setHtml($('#error-msg'), data.message).show();
```
Would you prefer this over replicating the selector?  Or do you like this broken out?  If broken out, would you use a variable, or just leave it as is?  I'm just trying to learn JavaScript code style preferences.

robrap commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -45,11 +43,11 @@ <h2 class="hd hd-2">Enrollment</h2>
-    </label>
+            ${_("Enter the reason why the students are to be manually enrolled or unenrolled.")}
+            ${_("This cannot be left blank and will be recorded and presented in Enrollment Reports.")}
+            ${_("Therefore, please give enough detail to account for this action.")}
```
It seems like we'd want one long string for translation purposes:
```
${_((
    "Enter the reason why the students are to be manually enrolled or unenrolled. "
    "This cannot be left blank and will be recorded and presented in Enrollment Reports. "
    "Therefore, please give enough detail to account for this action."
))}
```

cahrens replied:
Agreed

cahrens replied:
Though since this was existing code, I don't think @clrux has to fix it in this PR (there are similar bad patterns elsewhere in this file).

clrux replied:
Happy to fix since I'm in here anyway. Won't do every instance, but I'll get this one.

cahrens commented:
@clrux I'm getting 500 error on your sandbox on the dashboard page (https://clrux-ac-414.sandbox.edx.org/dashboard). I doubt this is related to your PR, but please ssh into the sandbox and check the LMS logs to verify the cause of the problem.

cahrens commented:
👍  Once sandbox is up and fully functional.

cahrens commented:
@clrux On the sandbox, it looks like Bulk Email and CCX are not yet enabled. 

I'm also seeing a console error that I don't recall seeing before (though the instructor dashboard unfortunately routinely throws many errors)--

instructor_dash.46c66a55c9f3.js:formatted:775 Uncaught TypeError: Cannot read property 'onClickTitle' of undefined

Can you please take a look and make sure this is not a new issue? The JavaScript looked like it could possibly be related.

````
sections_have_loaded.after(function() {
    return $section.data("wrapper").onClickTitle()
})

clrux commented:
@cahrens I'm having issues building the sandy. Let me investigate. I'll get back to you.

clrux commented:
@cahrens Looks like I got one that works! https://clrux-ac-414.sandbox.edx.org/courses/ccx-v1:edX+DemoX+Demo_Course+ccx@1/instructor#view-course_info

clrux commented:
@cahrens It looks like this error is on stage as well. It has something to do with the Analytics "tab"; it happens when you click it. This isn't caused by anything in this branch.

clrux commented:
@benpatterson Would you mind re-re-reviewing this one for me?

benpatterson commented:
@clrux I'm wondering if you have run the a11y tests after the recent changes to the JS assertion library, which you and @andy-armstrong put onto npm. The end-result should be the same, but it would be good to check on that.

Also, do you have ticket(s) logged for the wcag2aa TODOs that were added with this file? Sorry if I missed the reference.

cahrens commented:
@clrux Please also remember to squash all commits before merging.

clrux commented:
Updated the TODO tag to AC-491. Squashed. @benpatterson if you could formally thumbs, I'd appreciate it. I will follow up with you later today or Monday on your second question regarding the new a11y check output.

benpatterson commented:
@clrux My question earlier was not so much the check, but moreso "if you merge this to master, will it pass on master?"

I do think we need to test with the updated checks, and that that can come later. But I also want to be sure that when this merges, it'll be ok. Not sure if you've rebased recently, etc....

clrux commented:
@benpatterson This work takes into account all of the preexisting a11y checks we have on Master. I've ignored all `section` checks except for those on the Instructor Dashboard (the aim of this particular PR).

benpatterson commented:
thanks @clrux 

👍
